src/lang/builtins_files.ml

```diff
@@ -243,7 +241,20 @@ let () =
         if absolute then List.map (Filename.concat dir) files else files
       in
       let files = List.map Lang.string files in
-      Lang.list ~t:Lang.string_t files)
+      Lang.list files)
+
+let () =
+  add_builtin "file.metadata" ~cat:Sys
+    [("", Lang.string_t, None, Some "Read metadata from a file.")]
+    Lang.metadata_t
+    ~descr:"Call a function when a file is modified. Returns unwatch function."
+    (fun p ->
+      let uri = Lang.to_string (List.assoc "" p) in
+      let r = Request.create uri in
+      let ctype = { Frame.audio = 0; video = 0; midi = 0 } in
+      if Request.resolve ~ctype:(Some ctype) r 30. = Request.Resolved then
+        Lang.metadata (Request.get_all_metadata r)
+      else Lang.metadata (Hashtbl.create 0))
```

Comment:
Why was this moved back to builtin?

Reply:
Well done, you spotted a tricky point. Let me explain.

Because you don't determine the content type with typing anymore, I had to remove the type from requests: when you create you don't determine it content type. For instance, the type of `request.create` used to be
```
(...) -> request(audio='a, video='b, midi='c)
```
and is now
```
(...) -> request
```
Of course I did not remove the content type checks, but content type is determined at resolution time. For this reason, you cannot use `request.resolve` in Liquidsoap anymore, which explains why I had to move this to the OCaml side. All requests resolutions are performed in the OCaml side (not entirely true, see below).

Side note: there is still a `request.resolve` in Liquidsoap, but it resolves without content type: this corresponds to what you used to obtain with `request.create_raw` before and can only be used to download files which are not going to be decoded (e.g. playlists).

Reply:
Ho I see, `request.resolve` passes the negotiated content type down to the decoder at runtime. Cool, it was always tricky to explain how to properly resolve requests without blocking the main thread anyway.